pkg/controller/jobframework/reconciler.go

```diff
@@ -1148,24 +1150,62 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
-		conditionStatus = metav1.ConditionTrue
-		message = "All pods were ready or succeeded since the workload admission"
-	}
-	return metav1.Condition{
-		Type:    kueue.WorkloadPodsReady,
-		Status:  conditionStatus,
-		Reason:  "PodsReady",
-		Message: message,
-		// ObservedGeneration is added via workload.UpdateStatus
+func generatePodsReadyCondition(log logr.Logger, job GenericJob, wl *kueue.Workload) metav1.Condition {
+	if !workload.IsAdmitted(wl) {
+		// The workload has not been admitted yet
+		// or it was admitted in the past but it's evicted/requeued
+		return workload.CreatePodsReadyCondition(metav1.ConditionFalse,
+			kueue.WorkloadNotAdmitted,
+			"Workload isn't admitted")
+	}
+
+	podsReadyStartCond := workload.FindStatusConditionWithReason(wl, kueue.WorkloadPodsReady, kueue.WorkloadWaitForPodsStart)
+	podsReadyRecoveryCond := workload.FindStatusConditionWithReason(wl, kueue.WorkloadPodsReady, kueue.WorkloadWaitForPodsRecovery)
+	podsReady := job.PodsReady()
+	log.V(3).Info("Generating PodsReady condition",
+		"WaitForPodsReady with WorkloadWaitForPodsStart reason", podsReadyStartCond,
+		"WaitForPodsReady with WorkloadWaitForPodsRecovery reason", podsReadyRecoveryCond,
+		"Workload admitted", workload.IsAdmitted(wl),
+		"Pods are ready", podsReady)
+	switch {
+	// 1a. The workload has just been admitted and doesn't have PodsReady cond yet
+	// 1b. The workload has just been admitted, and not all pods are ready
+	case !podsReady && (apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadPodsReady) == nil ||
+		podsReadyStartCond != nil && podsReadyStartCond.Status == metav1.ConditionFalse):
+		return workload.CreatePodsReadyCondition(metav1.ConditionFalse,
+			kueue.WorkloadWaitForPodsStart,
+			"Not all pods are ready or succeeded")
+
+	// 2a. Pods have reached readiness; before job runtime
+	// 2b. Pods continue to be ready
+	case podsReady && podsReadyStartCond != nil:
+		return workload.CreatePodsReadyCondition(metav1.ConditionTrue,
+			kueue.WorkloadWaitForPodsStart,
+			"All pods reached readiness and the workload is running")
+
+	// 3a. The workload was running and a pod has failed
+	// 3b. A pod has failed and the Workload has already recovered once before
+	case !podsReady && (podsReadyStartCond != nil && podsReadyStartCond.Status == metav1.ConditionTrue ||
+		podsReadyRecoveryCond != nil):
+		return workload.CreatePodsReadyCondition(metav1.ConditionFalse,
+			kueue.WorkloadWaitForPodsRecovery,
+			"At least one pod has failed, waiting for recovery")
+
+	// 4a. A failed pod has recovered during workload's runtime
+	// 4b. Pods continue to be ready after recovery
+	case podsReady && podsReadyRecoveryCond != nil:
+		return workload.CreatePodsReadyCondition(metav1.ConditionTrue,
+			kueue.WorkloadWaitForPodsRecovery,
+			"All pods reached readiness and the workload is running")
+	default:
+		log.V(2).Info("WARNING Unexpected PodsReady condition",
```

Comment:
What is the assumed situations where we face here in this flow?
https://github.com/PBundyra/kueue/blob/daad59ea89bc520b9af7d3ed58c4117bf03ec337/keps/349-all-or-nothing/README.md

Reply:
Sorry for your confusion. This is just a question to clarify when we face this path (`default`).

Reply:
Now the `default` path is only for unexpected circumstances. Also, we need to return something if non of the above `cases` are true